common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,490 @@
+                                        mCurrentDialog.onCancelCba();
+                                        //Reset dialog to null
+                                        mCurrentDialog = null;
+                                        //TODO: maybe show a dialog that tells user they unplugged too early? I'm not sure this is necessary since I'm thinking in a lot of cases, the user wanted to cancel current flow by unplugging.
+                                        Logger.verbose(TAG, "YubiKey was disconnected while dialog was still displayed.");
+                                    }
+                                }
+                            }
+                        }
+                    });
+                }
             }
         });
     }
 
-    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    /**
+     * Called when a ClientCertRequest is received by the AzureActiveDirectoryWebViewClient.
+     * Prompts the user to choose a certificate to authenticate with based on whether or not a YubiKey is plugged in and has permission to be connected.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @return null in either case.
+     */
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
-        return handleOnDeviceCertAuth(request);
+        synchronized (deviceLock) {
+            if (mDevice != null) {
+                handleSmartcardCertAuth(request);
+                return null;
+            }
+        }
+        //Else, proceed with user certificates stored on device.
+        handleOnDeviceCertAuth(request);
+        return null;
+    }
+
+    /**
+     * Collects YubiKitCertDetails from PIV certificates on YubiKey.
+     * If certificates are found on YubiKey, the method proceeds with the smartcard certificate picker.
+     * Otherwise, appropriate error dialogs are shown.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":handleSmartcardCertAuth";
+        //A connection to the YubiKey needs to be made in order to read the certificates off it.
+        mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
+                try {
+                    final SmartCardConnection c = value.getValue();
+                    PivSession piv = new PivSession(c);
+                    //Check if too many PIN attempts have been made.
+                    if (piv.getPinAttempts() == 0) {
```

Comment:
So this is remaining pin attempts basically

Reply:
yes... are you asking about the comment? I can change it to  `//Check if there are no PIN attempts remaining.`

Reply:
I was referring to the method name i.e. `getPinAttempts`. I was hoping it would be named something like `getRemainingPinAttempts`. I believe this method is in YubyKit code so I don't think you can change this.